### ISSUE TITLE
BUG: kernel crashes on daskhub running cell no23 (actually running tractor)

### ISSUE THREAD

**Issue by bsipocz** [fornax-navo/fornax-demo-notebooks#47]
Running the smallest instance I keep running into kernel crashes. Haven't checked on the details yet, maybe it's a too small one?

Tasks here:
 - check the logs whether we see anything useful there (maybe Gerson, or we should document who and where to check those logs)
 - document at the top of the notebook the minimum system requirements.

**Comment by troyraen** [comment]
Just reporting that I encountered the same error on fornaxdev, smallest instance -- kernel crashed on cell 23, running tractor with multiprocessing.

**Comment by bsipocz** [comment]
I think this points us to the direction that it would be great to have two types of notebooks (or cells): one is for demonstration purposes, that can be run and that executes in a reasonably short time, while the other set (or a detailed narrative how to make the changes) will do the full use case, e.g run for a day to get all the photometry needed.

**Comment by troyraen** [comment]
Documenting some recent events:

I ran three tests on this multiprocessing cell that used different numbers of max workers (24, 12, and 6). As number of max workers goes down:
- memory usage decreases approx. linearly (~13.5, 7.4, and 4.2 GB respectively) — as expected.  
- wall time remains approx. constant (26min 54sec, 26min 6sec, and 25min 38sec respectively) — expected linear increase.

@bsipocz initial suspicion was that numpy multiprocessing has not been turned off. 

So the **_quick fix_** here is to lower the number of max workers being used, and then that cell runs just fine. So in the line

```
with concurrent.futures.ProcessPoolExecutor(24) as executor:
```

change `24` to something much smaller like `6`  (`1` might even work great, I just didn’t check).

No one has looked any further yet. A real fix still needs to be found.

**Comment by bsipocz** [comment]
I'm going ahead and close this issue as we fully explored the issue with this cell:

- crashes were due to running out of memory
- performance issues and differences between IRSA_SP and daskhub were due to differences is allocated CPUs, we didn't see speed improvements on daskhub for >4 as the available instances had only 4. #65 changed to code to use the default, max available, rather than a hard-wired number that maybe more than what's available on a given system.

**Comment by troyraen** [comment]
Thanks @bsipocz for tracking down the issue with CPUs > 4!

Just leaving two final notes here, after a little experimentation last night:

1. I confirm the performance gain (decreased processing time) you reported when scaling the `ProcessPoolExecutor`'s `max_workers` arg from 1-4, and no gain beyond 4. I was using a "Large Server", which has "2GB RAM/2 CPU reserved. Up to 8GB RAM/4 CPU as needed".
2. When leaving `max_workers` unset (so it defaults to whatever's available), memory usage was reasonable (i.e., within range of the capacity _reserved_ for my machine size), but only if I started out with a clean slate. If I started out with a bunch of stuff already loaded in memory, this setting would kick my memory usage all the way up to the limit available for bursting. I think the comment we left in that cell re: `max_workers` is clear enough for now. Just something to be aware of as we experiment with more multiprocessing.
